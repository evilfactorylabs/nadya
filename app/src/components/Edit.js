import React, { useState, useEffect } from 'react'
import DayjsUtils from '@date-io/dayjs'

import {
  Card,
  CardContent,
  Button,
  CardActions,
  TextField,
  InputLabel,
  InputAdornment,
  FormControl,
  OutlinedInput,
  MenuItem,
  Select
} from '@material-ui/core'
import { makeStyles } from '@material-ui/core/styles'
import { DatePicker, MuiPickersUtilsProvider } from '@material-ui/pickers'

const useStyles = makeStyles(theme => ({
  input: {
    margin: '0.5rem',
    marginTop: '1rem',
    marginBottom: '1rem'
  },
  action: {
    marginLeft: 'auto',
    marginRight: theme.spacing(1),
    marginBottom: theme.spacing(2)
  }
}))

export default ({ updateSubscription, subscription }) => {
  const classes = useStyles()
  const [serviceName, setServiceName] = useState('')
  const [cost, setCost] = useState('')
  const [costInterval, setCostInterval] = useState(0)
  const [startSubscription, setStartSubscription] = useState(new Date())

  const shouldButtonDisabled = () =>
    !serviceName || !cost || !costInterval || !startSubscription

  const handleUpdateSubscription = () => {
    const payload = {
      serviceName,
      costInterval,
      cost,
      firstBill: startSubscription
    }
    updateSubscription(payload)
  }

  useEffect(() => {
    setServiceName(subscription.title)
    setCost(subscription.cost)
    setCostInterval(subscription.period)
    setStartSubscription(subscription.firstBill)
  }, [subscription])

  return (
    <Card square={true}>
      <CardContent>
        <FormControl fullWidth>
          <TextField
            name='name'
            label='Nama layanan'
            margin='dense'
            variant='outlined'
            value={serviceName || ''}
            disabled={true}
            onChange={e => setServiceName(e.target.value)}
          />
        </FormControl>
        <FormControl fullWidth margin='normal'>
          <TextField
            name='cost'
            type='number'
            label='Biaya'
            margin='dense'
            variant='outlined'
            value={cost || 0}
            onChange={e => setCost(e.target.value)}
            InputProps={{
              startAdornment: (
                <InputAdornment position='start'>Rp </InputAdornment>
              )
            }}
          />
        </FormControl>
        <FormControl variant='outlined' fullWidth margin='dense'>
          <InputLabel variant='outlined'>Periode penagihan</InputLabel>
          <Select
            value={costInterval || 0}
            onChange={e => setCostInterval(e.target.value)}
            input={<OutlinedInput labelWidth={150} margin='none' />}
          >
            <MenuItem value={1}>Bulanan</MenuItem>
            <MenuItem value={2}>Tahunan</MenuItem>
          </Select>
        </FormControl>
        <MuiPickersUtilsProvider utils={DayjsUtils}>
          <DatePicker
            fullWidth
            margin='normal'
            label='Tanggal pertama berlangganan'
            inputVariant='outlined'
            value={startSubscription || new Date()}
            onChange={setStartSubscription}
          />
        </MuiPickersUtilsProvider>
      </CardContent>
      <CardActions>
        <Button
          className={classes.action}
          color='primary'
          variant='contained'
          disabled={shouldButtonDisabled()}
          onClick={handleUpdateSubscription}
        >
          Simpan
        </Button>
      </CardActions>
    </Card>
  )
}
