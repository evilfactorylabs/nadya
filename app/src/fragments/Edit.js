import React, { useState, useEffect } from 'react'
import DayjsUtils from '@date-io/dayjs'
import CardContent from '@material-ui/core/CardContent'
import Button from '@material-ui/core/Button'
import Card from '@material-ui/core/Card'
import CardActions from '@material-ui/core/CardActions'
import TextField from '@material-ui/core/TextField'
import InputLabel from '@material-ui/core/InputLabel'
import InputAdornment from '@material-ui/core/InputAdornment'
import FormControl from '@material-ui/core/FormControl'
import OutlinedInput from '@material-ui/core/OutlinedInput'
import MenuItem from '@material-ui/core/MenuItem'
import Select from '@material-ui/core/Select'
import { makeStyles } from '@material-ui/core/styles'
import { DatePicker, MuiPickersUtilsProvider } from '@material-ui/pickers'
import { getSubscription, updateSubscription } from 'services/subscription'
import { navigate } from '@reach/router'

const useStyles = makeStyles(theme => ({
  input: {
    margin: '0.5rem',
    marginTop: '1rem',
    marginBottom: '1rem'
  },
  action: {
    marginLeft: 'auto',
    marginRight: theme.spacing(1),
    marginBottom: theme.spacing(2)
  }
}))

export default ({ subscription_id }) => {
  const classes = useStyles()
  const [createdAt, setCreatedAt] = useState('')
  const [revId, setRevId] = useState('')
  const [serviceName, setServiceName] = useState('')
  const [cost, setCost] = useState('')
  const [costInterval, setCostInterval] = useState(0)
  const [startSubscription, setStartSubscription] = useState(new Date())

  const shouldButtonDisabled = () =>
    !serviceName || !cost || !costInterval || !startSubscription

  const handleUpdateSubscription = () => {
    const payload = {
      _id: subscription_id,
      _rev: revId,
      title: serviceName,
      period: costInterval,
      createdAt: createdAt,
      cost: cost,
      firstBill: new Date(startSubscription).getTime()
    }
    if (!payload._rev) navigate('/', { replace: true })
    updateSubscription(payload)
      .then(doc => {
        navigate(`/${doc.id}/`, { replace: true })
      })
      .catch(err => {
        throw err
      })
  }

  useEffect(() => {
    getSubscription(subscription_id)
      .then(doc => {
        if (!doc.title) {
          navigate('/', { replace: true })
        }
        setCreatedAt(doc.createdAt)
        setRevId(doc._rev)
        setCost(doc.cost)
        setCostInterval(doc.period)
        setStartSubscription(new Date(doc.firstBill))
        setServiceName(doc.title)
      })
      .catch(err => {
        throw err
      })
  }, [subscription_id])

  return (
    <Card square={true}>
      <CardContent>
        <FormControl fullWidth>
          <TextField
            name='name'
            label='Nama layanan'
            margin='dense'
            variant='outlined'
            value={serviceName}
            disabled={true}
            onChange={e => setServiceName(e.target.value)}
          />
        </FormControl>
        <FormControl fullWidth margin='normal'>
          <TextField
            name='cost'
            type='number'
            label='Biaya'
            margin='dense'
            variant='outlined'
            value={cost}
            onChange={e => setCost(e.target.value)}
            InputProps={{
              startAdornment: (
                <InputAdornment position='start'>Rp </InputAdornment>
              )
            }}
          />
        </FormControl>
        <FormControl variant='outlined' fullWidth margin='dense'>
          <InputLabel variant='outlined'>Periode penagihan</InputLabel>
          <Select
            value={costInterval}
            onChange={e => setCostInterval(e.target.value)}
            input={<OutlinedInput labelWidth={150} margin='none' />}
          >
            <MenuItem value={1}>Bulanan</MenuItem>
            <MenuItem value={2}>Tahunan</MenuItem>
          </Select>
        </FormControl>
        <MuiPickersUtilsProvider utils={DayjsUtils}>
          <DatePicker
            fullWidth
            margin='normal'
            label='Tanggal pertama berlangganan'
            inputVariant='outlined'
            value={startSubscription}
            onChange={setStartSubscription}
          />
        </MuiPickersUtilsProvider>
      </CardContent>
      <CardActions>
        <Button
          className={classes.action}
          color='primary'
          variant='contained'
          disabled={shouldButtonDisabled()}
          onClick={handleUpdateSubscription}
        >
          Simpan
        </Button>
      </CardActions>
    </Card>
  )
}
