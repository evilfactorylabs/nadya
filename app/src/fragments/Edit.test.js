import Edit from './Edit'

afterAll(async () => {
  await cleanup()
})

describe('Fragments: Edit.js', () => {
  test('It should render Edit screen', () => {
    const { container } = render(<Edit />)

    expect(container).toBeInTheDocument()
  })
})
