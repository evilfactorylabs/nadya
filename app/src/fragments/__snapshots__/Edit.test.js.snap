// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Fragments: Edit.js It should render Edit screen 1`] = `
<div>
  <div
    class="MuiPaper-root MuiPaper-elevation1 MuiCard-root"
  >
    <div
      class="MuiCardContent-root"
    >
      <div
        class="MuiFormControl-root MuiFormControl-fullWidth"
      >
        <div
          class="MuiFormControl-root MuiTextField-root MuiFormControl-marginDense"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-marginDense MuiInputLabel-outlined Mui-disabled Mui-disabled"
            data-shrink="false"
          >
            Nama layanan
          </label>
          <div
            class="MuiInputBase-root MuiOutlinedInput-root Mui-disabled Mui-disabled MuiInputBase-formControl MuiInputBase-marginDense MuiOutlinedInput-marginDense"
          >
            <fieldset
              aria-hidden="true"
              class="PrivateNotchedOutline-root-86 MuiOutlinedInput-notchedOutline"
              style="padding-left: 8px;"
            >
              <legend
                class="PrivateNotchedOutline-legend-87"
                style="width: 0.01px;"
              >
                <span>
                  ​
                </span>
              </legend>
            </fieldset>
            <input
              aria-invalid="false"
              class="MuiInputBase-input MuiOutlinedInput-input Mui-disabled Mui-disabled MuiInputBase-inputMarginDense MuiOutlinedInput-inputMarginDense"
              disabled=""
              name="name"
              type="text"
              value=""
            />
          </div>
        </div>
      </div>
      <div
        class="MuiFormControl-root MuiFormControl-marginNormal MuiFormControl-fullWidth"
      >
        <div
          class="MuiFormControl-root MuiTextField-root MuiFormControl-marginDense"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-marginDense MuiInputLabel-outlined MuiFormLabel-filled"
            data-shrink="true"
          >
            Biaya
          </label>
          <div
            class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl MuiInputBase-marginDense MuiOutlinedInput-marginDense MuiInputBase-adornedStart MuiOutlinedInput-adornedStart"
          >
            <fieldset
              aria-hidden="true"
              class="PrivateNotchedOutline-root-86 MuiOutlinedInput-notchedOutline"
              style="padding-left: 8px;"
            >
              <legend
                class="PrivateNotchedOutline-legend-87"
                style="width: 0px;"
              >
                <span>
                  ​
                </span>
              </legend>
            </fieldset>
            <div
              class="MuiInputAdornment-root MuiInputAdornment-positionStart"
            >
              <p
                class="MuiTypography-root MuiTypography-body1 MuiTypography-colorTextSecondary"
              >
                Rp 
              </p>
            </div>
            <input
              aria-invalid="false"
              class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMarginDense MuiOutlinedInput-inputMarginDense MuiInputBase-inputAdornedStart MuiOutlinedInput-inputAdornedStart"
              name="cost"
              type="number"
              value="0"
            />
          </div>
        </div>
      </div>
      <div
        class="MuiFormControl-root MuiFormControl-marginDense MuiFormControl-fullWidth"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-marginDense MuiInputLabel-outlined MuiFormLabel-filled"
          data-shrink="true"
        >
          Periode penagihan
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl"
        >
          <fieldset
            aria-hidden="true"
            class="PrivateNotchedOutline-root-86 MuiOutlinedInput-notchedOutline"
            style="padding-left: 8px;"
          >
            <legend
              class="PrivateNotchedOutline-legend-87"
              style="width: 120.5px;"
            >
              <span>
                ​
              </span>
            </legend>
          </fieldset>
          <div
            class="MuiSelect-root"
          >
            <div
              aria-haspopup="true"
              aria-pressed="false"
              class="MuiSelect-select MuiSelect-selectMenu MuiInputBase-input MuiOutlinedInput-input MuiSelect-outlined"
              role="button"
              tabindex="0"
            />
            <input
              type="hidden"
              value="0"
            />
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSelect-icon"
              focusable="false"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M7 10l5 5 5-5z"
              />
            </svg>
          </div>
        </div>
      </div>
      <div
        class="MuiFormControl-root MuiTextField-root MuiFormControl-marginNormal MuiFormControl-fullWidth"
        format="MMMM Do"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-filled"
          data-shrink="true"
        >
          Tanggal pertama berlangganan
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl"
        >
          <fieldset
            aria-hidden="true"
            class="PrivateNotchedOutline-root-86 MuiOutlinedInput-notchedOutline"
            style="padding-left: 8px;"
          >
            <legend
              class="PrivateNotchedOutline-legend-87"
              style="width: 0px;"
            >
              <span>
                ​
              </span>
            </legend>
          </fieldset>
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiOutlinedInput-input"
            readonly=""
            type="text"
            value="July 2nd"
          />
        </div>
        
      </div>
    </div>
    <div
      class="MuiCardActions-root MuiCardActions-spacing"
    >
      <button
        class="MuiButtonBase-root MuiButton-root makeStyles-action-2 MuiButton-contained MuiButton-containedPrimary Mui-disabled Mui-disabled"
        disabled=""
        tabindex="-1"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          Simpan
        </span>
      </button>
    </div>
  </div>
</div>
`;
